contracts/zero-ex/contracts/src/features/MultiplexFeature.sol

```diff
@@ -0,0 +1,799 @@
+            "MultiplexFeature::multiHopFill/MISMATCHED_ARRAY_LENGTHS"
+        );
+        IERC20TokenV06 outputToken = IERC20TokenV06(fillData.tokens[fillData.tokens.length - 1]);
+        uint256 senderBalanceBefore = outputToken.getTokenBalanceOf(msg.sender);
+        uint256 ethBalanceBefore = address(this).balance.safeSub(msg.value);
+
+        _multiHopFill(fillData, msg.value);
+
+        // RFC: I think these balance checks are arguably unnecessary,
+        // see reasoning above.
+
+        // The `outputTokenAmount` returned by `_multiHopFill` may not
+        // be fully accurate (e.g. due to some janky token and/or
+        // reentrancy situation).
+        outputTokenAmount = outputToken.getTokenBalanceOf(msg.sender)
+            .safeSub(senderBalanceBefore);
+        require(
+            outputTokenAmount >= minBuyAmount,
+            "MultiplexFeature::multiHopFill/UNDERBOUGHT"
+        );
+
+        uint256 ethBalanceAfter = address(this).balance;
+        require(
+            ethBalanceAfter >= ethBalanceBefore,
+            "MultiplexFeature::multiHopFill/OVERSPENT_ETH"
+        );
+        // Refund ETH
+        if (ethBalanceAfter > ethBalanceBefore) {
+            payable(msg.sender).transfer(ethBalanceAfter - ethBalanceBefore);
+        }
+    }
+
+    // Similar to FQT. If `fillData.sellAmount` is set to `type(uint256).max`,
+    // this is effectively a batch fill. Otherwise it can be set to perform a
+    // market sell of some amount. In the case of a market sell, the `sellAmount`
+    // values of the constituent wrapped calls can encode percentages of the
+    // full `fillData.sellAmount`. This is useful if e.g. the `_batchFill` call
+    // the second leg of a multi-hop fill, where the sell amount is not known
+    // ahead of time.
+    function _batchFill(BatchFillData memory fillData, uint256 totalEth)
+        public
+        returns (uint256 outputTokenAmount, uint256 remainingEth)
+    {
+        // Track the remaining ETH allocated to this call.
+        remainingEth = totalEth;
+        // Track the amount of input token sold.
+        uint256 soldAmount;
+        for (uint256 i = 0; i != fillData.calls.length; i++) {
+            // Check if we've hit our target.
+            if (soldAmount >= fillData.sellAmount) { break; }
+            WrappedBatchCall memory wrappedCall = fillData.calls[i];
+            // Compute the fill amount.
+            uint256 inputTokenAmount = _normalizeAmount(
+                wrappedCall.sellAmount,
+                fillData.sellAmount,
+                fillData.sellAmount.safeSub(soldAmount)
+            );
+            if (wrappedCall.selector == INativeOrdersFeature._fillRfqOrder.selector) {
+                // Decode the RFQ order and signature.
+                (
+                    LibNativeOrder.RfqOrder memory order,
+                    LibSignature.Signature memory signature
+                ) = abi.decode(
```

Comment:
I imagine this will get pretty expensive. For each RFQ order we decode / encode / decode. I think it would save a lot of gas if we just had the single decode in NativeOrdersFeature. 

Manual memory management does come with risks and added complexity. Thinking a bit on this. One way to sidestep could be encoding `wrappedCall.data` as the input to the NativeOrdersFeature: `(Order, Signature, Amount, Sender)`. Then we simply fill in the Amount/Sender here execute the call.

Reply:
Looks like the gas savings are already substantial so probably not worth further optimizing at this time. But, something to consider going forward.